protocol/contracts/v2/TreasuryReservesVault.sol

```diff
@@ -0,0 +1,474 @@
+        if (!strategyData.isEnabled) revert NotEnabled();
+        if (strategyData.isShuttingDown) revert StrategyIsShutdown();
+
+        emit Borrow(strategyAddr, recipient, borrowAmount);
+
+        // Mint new dUSD, source stables from the baseStrategy and send the strategy wanting to borrow.
+        internalDebtToken.mint(strategyAddr, borrowAmount);
+        _withdrawAndSendStables(recipient, borrowAmount);
+    }
+    
+    function _withdrawAndSendStables(address to, uint256 amount) internal {
+        // If a baseStrategy is defined, then pull as many stables directly from this
+        // contract (not yet deposited into the baseStrategy)
+        // and then pull the rest from baseStrategyDeposits.
+        ITempleBaseStrategy _baseStrategy = baseStrategy;
+        if (address(_baseStrategy) != address(0)) {
+            uint256 withdrawFromBaseStrategyAmount;
+            {
+                uint256 _balance = stableToken.balanceOf(address(this));
+                uint256 _directTransferAmount = _balance > amount ? amount : _balance;
+                withdrawFromBaseStrategyAmount = amount - _directTransferAmount;
+            }
+
+            // Pull any remainder required from the base strategy.
+            // This will ensure we get at least the min threshold amount.
+            if (withdrawFromBaseStrategyAmount != 0) {
+                // So there aren't many small withdrawals, ensure we withdraw at least
+                // some minimum threshold amount each time.
+                uint256 _threshold = baseStrategyWithdrawalThreshold;
+                uint256 _stablesToWithdraw = withdrawFromBaseStrategyAmount < _threshold ? _threshold : withdrawFromBaseStrategyAmount;
+
+                // The amount actually withdrawn may be less than requested
+                // as it's capped to any actual remaining balance.
+                uint256 _withdrawnAmount = _baseStrategy.trvWithdraw(_stablesToWithdraw);
+
+                // Burn that amount of dUSD from the base strategy.
+                internalDebtToken.burn(address(_baseStrategy), _withdrawnAmount);
+            }
+        }
+
+        // Finally send the stables.
+        stableToken.safeTransfer(to, amount);
+    }
+
+    // @todo add a repay function where TRV/RAMOS can pay down some of the debt with Temple.
+    // This will burn the Temple, and burn the equivalent amount of dUSD
+    function repayTemple(uint256 repayAmount, address strategy) external override {
+        if (repayAmount == 0) revert CommonEventsAndErrors.ExpectedNonZero();
+        if (globalRepaysPaused) revert RepaysPaused();
+
+        Strategy storage strategyData = strategies[strategy];
+        if (!strategyData.isEnabled) revert NotEnabled();
+        if (strategyData.repaysPaused) revert RepaysPaused();
+
+        // dUSD price = Temple * TPI
+        uint256 debtToBurn = repayAmount * treasuryPriceIndex() / 1e18;
+
+        // @todo need an event like:?
```

Comment:
could also add token address as argument to `Repay` event to differentiate tokens

Reply:
Yes...so there was a change in spec, as is the temple way! I didn't want to pollute your current PR though

I'm currently rejigging the TRV such that it can be configured to borrow/repay any configured token (and debt token pair).

I'll give you details in Discord.

Reply:
learned a new word, rejigging lol